### PR TITLE
Compile virtual dependencies before finalizing stubs

### LINKED ISSUE #2034: Missing overloads due to virtual stub finalization order
Concrete class discovery when finalizing virtual stubs still has an ordering issue on cross dependencies (as it seems, we recently fixed chained dependencies), leading to some overloads to be missed in certain cases, in turn running into an assertion when the virtual stub is called. One situation I have seen looks roughly like the following (probably not 100% accurate), but only runs into issues depending on compilation and finalization order:

```ts
class Base {
  abstract foo();
  abstract bar();
}

class A extends Base {
  foo() { new B().foo(); }
  bar() { ... }
}

class B extends Base {
  foo() { new A().bar(); }
  bar() { ... }
}

// somewhere deep:
someBase.foo(); // ?
```

A potential fix could be to add a separate discovery phase before finalizing.

cc @Scutterman

### PR DESCRIPTION
**Pull request by dcodeIO**
Should fix #2034 by first compiling virtual overload dependencies so late cross-dependencies are not missed during stub finalization.

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

File: src/compiler.ts [hunk 1]
```diff
@@ -6991,102 +7003,69 @@ export class Compiler extends DiagnosticEmitter {
...
+        );
+      }
+      let classInstance = assert(overloadInstance.getClassOrInterface());
+      builder.addCase(classInstance.id, stmts);
+      // Also alias each extendee inheriting this exact overload
+      let extendees = classInstance.getAllExtendees(
+        overloadInstance.isAccessor
+          // FIXME: Parent is not the property, but should be, so we don't need
+          // to use a substring here but can trivially use the property name.
+          ? instance.prototype.name.substring(GETTER_PREFIX.length)
```
Comment (dcodeIO) [inline src/compiler.ts:7063]: This is a little odd, in that resolving a getter or setter makes their instance reference the class as the parent (as natural when resolving methods), not the property instance (which is merely an ephemeral pre-resolve container). All we need here, however, is the name of the property, which we cannot trivially obtain due to the strange parent reference, so we end up with some unnecessary string garbage.

**Comment by MaxGraey** [comment]
Any test which test this issue?

**Comment by dcodeIO** [comment]
This one is ordering dependent and I haven't yet managed to create a reliable reproduction. Ordering changes in the affected fixtures however indicate that the mechanism fixes ordering.

**Comment by MaxGraey** [comment]
I see.

**Comment by Scutterman** [comment]
There's a "minimal" reproduction here - about 8 typescript files.
https://github.com/Scutterman/assemblyscript-virtual-bug

**Comment by dcodeIO** [comment]
Giving the repro a first try, the fix appears to do what it is supposed to

```
discover assembly/actions/lemmingAction/LemmingAction#label
  assembly/actions/fall/Fall#label, kind=5
discover assembly/actions/lemmingAction/LemmingAction#update
  assembly/actions/fall/Fall#update, kind=5
finalize assembly/actions/lemmingAction/LemmingAction#label
  assembly/actions/fall/Fall#label
  assembly/actions/walk/Walk#label
finalize assembly/actions/lemmingAction/LemmingAction#update
  assembly/actions/fall/Fall#update
  assembly/actions/walk/Walk#update
```

but the call still produces an `unreachable`.

**Comment by dcodeIO** [comment]
Turns out that issue with cross-dependencies runs even deeper. Was able to fix the remaining problem in an ugly way, but feels like there must be a better solution.

**Comment by MaxGraey** [comment]
> Turns out that issue with cross-dependencies runs even deeper.

What does it mean? It's problem with cyclic class dependencies?

**Comment by dcodeIO** [comment]
There are multiple levels of dependencies that can be discovered, and the fix right now isn't catching all of them. Needs some sort of bookkeeping over multiple loops it seems.

**Comment by dcodeIO** [comment]
Now loops until no more virtual calls *or* overloads are discovered, which fixes the repro.

**Pull request merged**